### DISCUSSION TITLE
Can't connect to a speaker. How to figure out what's wrong?

### DISCUSSION THREAD

**Discussion by kchudoba** [pschatzmann/ESP32-A2DP#500] [Q&A] [answered]
So I tested several examples using this library and also the arduino-audio-tools library. My ESP32-CAM connects without issues to my Sony headphones, but I can't make it connect to my standalone speaker. When I don't provide the device name to connect, even after flashing and restarting, the ESP32 doesn't even try to connect to the speaker (it doesn't show anything). Putting the name of the speaker doesn't help either. My iPhone and Win10 laptop connect to the speaker, even when I remove/forget the BT device in the settings of said iphone/laptop (which I assume means that it is still in pairing mode?). 

Is there a specific BT version required by the ESP that my speaker doesn't support? 
How can I debug it?
Any settings in the A2DP library that I can try out to see if I get a connection?
Any other sketches I can try that would perhaps list all visible BT devices?
Any other ideas?

Thanks in advance!

**Comment by kchudoba** [comment]
the debug log shows the following:

[ 36773][D][BluetoothA2DPSource.cpp:460] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 0
[ 36773][I][BluetoothA2DPSource.cpp:393] filter_inquiry_scan_result(): [BT_AV] Scanned device: **fc:58:fa:55:f9:6d**
[ 36780][I][BluetoothA2DPSource.cpp:399] filter_inquiry_scan_result(): [BT_AV] --Class of Device: 0x260404
[ 36790][I][BluetoothA2DPSource.cpp:403] filter_inquiry_scan_result(): [BT_AV] --RSSI: -80
[ 40122][D][BluetoothA2DPSource.cpp:460] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 0
[ 40123][I][BluetoothA2DPSource.cpp:393] filter_inquiry_scan_result(): [BT_AV] Scanned device: ac:9b:0a:4f:d7:a1
[ 40130][I][BluetoothA2DPSource.cpp:399] filter_inquiry_scan_result(): [BT_AV] --Class of Device: 0x240404
[ 40139][I][BluetoothA2DPSource.cpp:403] filter_inquiry_scan_result(): [BT_AV] --RSSI: -68
[ 40147][I][BluetoothA2DPSource.cpp:423] filter_inquiry_scan_result(): [BT_AV] --Compatiblity: Compatible
[ 40156][I][BluetoothA2DPSource.cpp:425] filter_inquiry_scan_result(): [BT_AV] --Name: MDR-ZX770BN
available SSID: MDR-ZX770BN


the first device is the BT speaker. It doesn't show a name (even though it does appear in windows/ios etc). Also the compatibility debug line is missing. The CoD is also different. 
@pschatzmann do you think there is a way to make it work or is the speaker simply not compatible with the ESP32?

**Comment by pschatzmann** [comment, marked as answer]
Why don't you just check in the source code. The log even tells you the line that you should look at.
Obviously your device does not provide any eir data, so the name can't be looked up....

**Reply by kchudoba** [reply]
yes I was just looking at it now. I am just a beginner and thus it's not so easy for me to understand the code. Is the name necessary to connect? Can't the connection be made using the bda? And how come other devices can see the name?

**Comment by pschatzmann** [comment]
I suggest you double check if you receive the name with ESP_BT_GAP_DEV_PROP_BDNAME and extend / adjust the logic to cover your case as well.

**Reply by kchudoba** [reply]
seems like I only get 2 props for the speaker - the COD and RSSI (i logged the param->disc_res.num_prop value). Could it be that the name is stored in a different profile (A2DP not GAT)? 
Would it be still possible to connect to it with the MAC address (and if yes, can you point me in the direction where the connection is actually established in the code)?

**Comment by pschatzmann** [comment]
Just look at the class documentation! 
You can call start w/o parameter and use the mentioned callback to check the ssid
There is even an [example](https://github.com/pschatzmann/ESP32-A2DP/blob/main/examples/bt_music_sender_ssid_callback/bt_music_sender_ssid_callback.ino)!
